### PR TITLE
feat: fix linear pricing migration

### PR DESCRIPTION
**Pull request by wischli**
# Description

* Bumps loans storage version from 3 to 4
* Adds migration for new `with_linear_pricing` handle of `ExternalPrice` struct
* To be checked: Sets `with_linear_pricing` true for all active (8 on mainnet) and created (1 on mainnet) loans

Please ignore temporary adjustments to tomls and the added logs, they are purely for debugging and will be removed!

## TODO

- [x] Fix arithmetic underflow error occurring during the nav update of the first asset already: When we call `y_coord_in_rect` in `maybe_with_linear_accrual_price`, we seem to have infeasible values:

* `oracle_provided_at == price_last_updated == x1 = 1715869380`
* `price == y1 = 0`
* `maturity == x2 = 1715212800`
* `self.info.notional == y2 = 100000000`
* `T::Time::now() == x = 1715949888`

Since `x1 > x2`, `y_coord_in_rect` performs `x1.ensure_sub(x)?` which throws because `now` is larger than the oracle timestamp `x1`. Logically, I don't think this could ever be right because the oracle timestamp should always be in the past?! So I imagine the order of the arguments of the `y_coord_in_rect` call is wrong here.

Logs
```
[2024-05-17T13:52:51Z WARN  pallet_loans::entities::pricing::external] current_price_inner: Seconds 1715212800, oracle Some((0, 1715869380568))
[2024-05-17T13:52:51Z WARN  pallet_loans::entities::pricing::external] maybe_with_linear_accrual_price: Seconds 1715212800, price 0, price_last_updated 1715869380
[2024-05-17T13:52:51Z WARN  cfg_utils::math] y_coord_in_rect: x1 1715869380, x2 1715212800, y1 0, y2 100000000, x: 1715949888
[2024-05-17T13:52:51Z ERROR runtime_common::migrations::loans] LoansMigrationToV4: error updating the portfolio for 4139607887: Arithmetic(Underflow)
```

Fixes #(issue)

## Changes and Descriptions

[List your changes here]

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: runtime/common/src/migrations/loans.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	pub type ActiveLoans<T: Config> = StorageMap<
+		Loans<T>,
+		Blake2_128Concat,
+		<T as Config>::PoolId,
+		ActiveLoansVec<T>,
+		ValueQuery,
+	>;
+
+	#[storage_alias]
+	pub type CreatedLoan<T: Config> = StorageDoubleMap<
```
Comment (lemunozm) [inline runtime/common/src/migrations/loans.rs:51]: Just curious, are we sure we do not need to migrate `ClosedLoan`? Will the expected decoded vector with that empty value match the current structure?

File: pallets/loans/src/entities/pricing/external.rs [hunk 2]
```diff
@@ -318,3 +320,110 @@ impl<T: Config> ExternalActivePricing<T> {
...
+	}
+
+	impl<T: Config> ExternalActivePricing<T> {
+		pub fn migrate(self, with_linear_pricing: bool) -> super::ExternalActivePricing<T> {
+			super::ExternalActivePricing {
+				info: self.info.migrate(with_linear_pricing),
+				outstanding_quantity: self.outstanding_quantity,
+				interest: self.interest,
+				latest_settlement_price: self.latest_settlement_price,
+				settlement_price_updated: self.settlement_price_updated,
```
Comment (lemunozm) [inline pallets/loans/src/entities/pricing/external.rs:410]: We need to ensure here that `settlement_price_updated` does no exceed the maturity date. [ref](https://github.com/centrifuge/centrifuge-chain/pull/1842/files#diff-637356fbd7809c17265d6b8ab519abc2b0ffa4df672cfb33a321795190f441baR437)

File: runtime/common/src/migrations/loans.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+			let mut changed_pools = Vec::new();
+			pallet_loans::ActiveLoans::<T>::translate::<v3::ActiveLoansVec<T>, _>(
+				|pool_id, active_loans| {
+					changed_pools.push(pool_id);
+					weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+					Some(
+						active_loans
+							.into_iter()
+							.map(|(loan_id, active_loan)| (loan_id, active_loan.migrate(true)))
```
Comment (lemunozm) [inline runtime/common/src/migrations/loans.rs:105]: Are we sure we want all loans `with_linear_accrual == true`? cc @mustermeiszer

File: pallets/loans/src/entities/pricing/external.rs (hunk 2 above)
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:410]: Never mind, after the last fix update. You do not need to do anything in this PR regarding this.

File: runtime/common/src/migrations/loans.rs (hunk 3 above)
Comment (wischli) [reply runtime/common/src/migrations/loans.rs:105]: Yes please double check this. The implementation is based on [this comment](https://kflabs.slack.com/archives/C04GJQAM9P0/p1715871507753699?thread_ts=1715849878.667849&cid=C04GJQAM9P0) from @hieronx.

File: runtime/common/src/migrations/loans.rs (hunk 1 above)
Comment (wischli) [reply runtime/common/src/migrations/loans.rs:51]: What I wanted to say: We only need to migrate if this storage is not empty. Right now, it is empty and it seems reasonable we can keep this state until upgrading Mainnet.

However, then I checked Demo and it has one entry. So yes, I will write a migration!

File: runtime/common/src/migrations/loans.rs (hunk 3 above)
Comment (hieronx) [reply runtime/common/src/migrations/loans.rs:105]: Just confirmed again with Anemoy, yes they want linear pricing to be on for all existing assets.

File: runtime/common/src/migrations/loans.rs (hunk 1 above)
Comment (lemunozm) [reply runtime/common/src/migrations/loans.rs:51]: Ah ok, I didn't know it was empty in centrifuge

**Comment by wischli** [comment]
@lemunozm There still seems to be an issue. We have surpassed the underflow error and now experience a `DivisionByZero` one:
```
[2024-05-23T09:47:41Z WARN  pallet_loans::entities::pricing::external] current_price_inner: Seconds 1715212800, oracle Some((0, 1715869380568))
[2024-05-23T09:47:41Z WARN  pallet_loans::entities::pricing::external] maybe_with_linear_accrual_price: maturity 1715212800, price 0, price_last_updated 1715869380
[2024-05-23T09:47:41Z WARN  cfg_utils::math] y_coord_in_rect: x1 1715212800, x2 1715212800, y1 0, y2 100000000, x: 1715212800
[2024-05-23T09:47:41Z ERROR runtime_common::migrations::loans] LoansMigrationToV4: error updating the portfolio for 4139607887: Arithmetic(DivisionByZero)
```

**Comment by lemunozm** [comment]
@wischli great catch! It should be fixed now after you rebase the branch again. I've added more testing to hope checking everything... 😅

File: runtime/common/src/migrations/loans.rs (hunk 1 above)
Comment (wischli) [reply runtime/common/src/migrations/loans.rs:55]: `ClosedLoan` migration added here: [833b9d1](https://github.com/centrifuge/centrifuge-chain/pull/1838/commits/833b9d125a15b4bb3328f60fe17f51ffebfe30d5)

**Review by lemunozm — APPROVED** [review]
Thanks for your super migration on this!! Suuuper clean!

**Pull request merged**